internal/config/config.go

```diff
@@ -302,6 +401,88 @@ func parsePeer(p peer) (*Peer, error) {
 	}, nil
 }
 
+// Parse peer autodiscovery configuration.
+//
+// BGP peer configuration can be specified in either annotations or labels. If
+// the same parameter is specified using both annotations and labels, the value
+// specified in annotations is used. If a parameter isn't specified in
+// annotations nor in labels, a default value - if configured - is used.
+//
+// If a required BGP parameter isn't specified in annotations, labels or
+// defaults, an error is returned. Required parameters:
+// - Local ASN
+// - Peer ASN
+// - Peer address
+func parsePeerAutodiscovery(p peerAutodiscovery) (*PeerAutodiscovery, error) {
+	pad := &PeerAutodiscovery{}
+
+	if p.FromAnnotations == nil && p.FromLabels == nil {
+		return nil, errors.New("missing from-annotations or from-labels")
+	}
+
+	if p.Defaults != nil {
+		pad.Defaults = &PeerAutodiscoveryDefaults{
+			ASN:     p.Defaults.ASN,
+			MyASN:   p.Defaults.MyASN,
+			Address: net.ParseIP(p.Defaults.Address),
+			Port:    p.Defaults.Port,
+		}
+
+		if p.Defaults.HoldTime != "" {
+			ht, err := parseHoldTime(p.Defaults.HoldTime)
+			if err != nil {
+				return nil, fmt.Errorf("parsing default hold time: %s", err)
+			}
+			pad.Defaults.HoldTime = ht
+		}
+	}
+
+	if p.FromAnnotations != nil {
+		pad.FromAnnotations = &PeerAutodiscoveryMapping{
+			ASN:      p.FromAnnotations.ASN,
+			Addr:     p.FromAnnotations.Addr,
+			HoldTime: p.FromAnnotations.HoldTime,
+			MyASN:    p.FromAnnotations.MyASN,
+			Port:     p.FromAnnotations.Port,
+			RouterID: p.FromAnnotations.RouterID,
+		}
+	}
+
+	if p.FromLabels != nil {
+		pad.FromLabels = &PeerAutodiscoveryMapping{
+			ASN:      p.FromLabels.ASN,
+			Addr:     p.FromLabels.Addr,
+			HoldTime: p.FromLabels.HoldTime,
+			MyASN:    p.FromLabels.MyASN,
+			Port:     p.FromLabels.Port,
+			RouterID: p.FromLabels.RouterID,
+		}
+	}
+
+	// We use a non-pointer in the raw json object, so that if the
+	// user doesn't provide a node selector, we end up with an empty,
+	// but non-nil selector, which means "select everything".
```

Comment:
It's not outdated, but I guess the comment is confusing. The raw JSON object mentioned is the following:

https://github.com/kinvolk/metallb/blob/d64bcb885c6d713469382f3fe82f5293eb23a592/internal/config/config.go#L47

I think you're right, even if we did use a pointer and it were nil, we could have a nil check instead of the length check and end up with the same behavior.

I've actually copied this part of the code from here:

https://github.com/kinvolk/metallb/blob/d64bcb885c6d713469382f3fe82f5293eb23a592/internal/config/config.go#L379-L393

I guess I can look into keeping things DRY here. As for the comment, should I change it?

Reply:
Refactored into a new function and removed the confusing comment.